src/runtime/vm/vm.cc

```diff
@@ -0,0 +1,660 @@
+  for (size_t i = 0; i < vm_func.instructions.size(); ++i) {
+    os << i << ": ";
+    InstructionPrint(os, vm_func.instructions[i]);
+    os << ";" << std::endl;
+  }
+}
+
+std::ostream& operator<<(std::ostream& os, const VMFunction& vm_func) {
+  VMFunctionPrint(os, vm_func);
+  return os;
+}
+
+void VirtualMachine::PushFrame(Index arg_count, Index ret_pc, const VMFunction& vm_func) {
+  auto frame = VMFrame(ret_pc, func_index, arg_count, code, vm_func.register_file_size);
+  frames.push_back(frame);
+}
+
+Index VirtualMachine::PopFrame() {
+  CHECK_NE(frames.size(), 0);
+  const VMFrame& fr = frames.back();
+  func_index = fr.func_index;
+  code = fr.code;
+  pc = fr.pc;
+  auto call_stack_size = frames.size();
+  frames.pop_back();
+  return call_stack_size;
+}
+
+void VirtualMachine::InvokeGlobal(const VMFunction& func, const std::vector<Object>& args) {
+  DLOG(INFO) << "===================\nInvoking global " << func.name << " " << args.size()
+                  << std::endl;
+
+  PushFrame(func.params, this->pc + 1, func);
+  for (size_t i = 0; i < args.size(); ++i) {
+    WriteRegister(i, args[i]);
+  }
+  DLOG(INFO) << "func.params= " << func.params << std::endl;
+
+  code = func.instructions.data();
+  pc = 0;
+}
+
+Object VirtualMachine::Invoke(const VMFunction& func, const std::vector<Object>& args) {
+  DLOG(INFO) << "Executing Function: " << std::endl << func << std::endl;
+
+  InvokeGlobal(func, args);
+  Run();
+  auto alloc = MemoryManager::Global()->GetAllocator(ctxs[0]);
+  DLOG(INFO) << "Memory used: " << alloc->UsedMemory() << " B\n";
+  return return_register;
+}
+
+Object VirtualMachine::Invoke(const std::string& name, const std::vector<Object>& args) {
+  auto func_index = this->global_map_[name];
+  DLOG(INFO) << "Invoke Global " << name << " at index " << func_index << std::endl;
+  return Invoke(this->functions[func_index], args);
+}
+
+void InvokePacked(const PackedFunc& func, Index arg_count, Index output_size,
+                  const std::vector<Object>& args) {
+  std::vector<TVMValue> values(arg_count);
+  std::vector<int> codes(arg_count);
+  runtime::TVMArgsSetter setter(values.data(), codes.data());
+
+  for (Index i = 0; i < arg_count; i++) {
+    NDArray data = ToNDArray(args[i]);
+    setter(i, data);
+  }
+
+  TVMRetValue rv;
+  func.CallPacked(TVMArgs(values.data(), codes.data(), arg_count), &rv);
+}
+
+void VirtualMachine::Init(const std::vector<TVMContext>& ctxs) { this->ctxs = ctxs; }
+
+inline void VirtualMachine::WriteRegister(Index r, const Object& val) {
+  frames.back().register_file[r] = val;
+}
+
+inline Object VirtualMachine::ReadRegister(Index r) const {
+  return frames.back().register_file[r];
+}
+
+void VirtualMachine::Run() {
+  CHECK(this->code);
+  this->pc = 0;
+  Index frame_start = frames.size();
+  while (true) {
+  main_loop:
+    auto const& instr = this->code[this->pc];
+    DLOG(INFO) << "\nExecuting(" << pc << "): ";
+#if USE_RELAY_LOG
```

Comment:
This flag does not have a switch in build file.

Reply:
This now should be `USE_RELAY_DEBUG` since we've removed the verbose control.

Reply:
@jroesch use `USE_RELAY_DEBUG` here?